Annotate framework assemblies as trimmable

sbomer opened the pull request:
This will opt them into trimming with the behavior described at https://github.com/mono/linker/blob/main/docs/design/trimmed-assemblies.md#assemblymetadataistrimmable-true.

The linker part of this change is happening in https://github.com/mono/linker/pull/1839.

@marek-safar @eerhardt @vitek-karas

sbomer committed:
Annotate framework assemblies as trimmable

This will opt them into trimming with the behavior described at https://github.com/mono/linker/blob/main/docs/design/trimmed-assemblies.md#assemblymetadataistrimmable-true.

eerhardt commented on eng/versioning.targets:
```diff
@@ -20,6 +20,9 @@
     <AssemblyMetadata Include="PreferInbox">
       <Value>True</Value>
     </AssemblyMetadata>
+    <AssemblyMetadata Include="IsTrimmable">
```
Are we sure we want to do this blindly? It feels like at least there should be a Condition to allow individual assemblies to turn it off.

Note that today we only analyze the assemblies in the shared fx / runtimepack for ILLink warnings. All OOB assemblies go unanalyzed. Once we have a story for analyzing libraries, we should hook `dotnet/runtime` up to it.

sbomer replied:
I can add a condition. Are there any in particular that you think I should turn off to start with? (Would it maybe make sense to turn it off for OOB assemblies and opt them in individually when we analyze them?)

sbomer replied:
Also, note that the feature as designed isn't tied to the warnings inherently - the attribute just says "trim me" - whether the assembly is linker-safe or not. That said, maybe it makes sense to opt out of trimming in cases where we don't trim a library during the runtime build because it uses too much reflection, like this:
https://github.com/dotnet/runtime/blob/75d5cc3df4dde3b37d75f289b5a90e358da838aa/src/libraries/System.Private.DataContractSerialization/src/System.Private.DataContractSerialization.csproj#L7-L8

marek-safar replied:
All libraries need to be trimmable in 6.0 but I'd probably exclude netfx build targets

eerhardt replied:
> All libraries need to be trimmable in 6.0

I agree we want to get there. But I think we should do it in a pragmatic way. Turning it on for all assemblies and letting people tell us what breaks doesn't seem like the best approach.

My thinking is that once we are done annotating the shared fx libraries, we start analyzing ILLink warnings for OOB libraries. When each library is "clean" of ILLink warnings, we mark it as "Trimmable".

For the initial change, I was assuming we would analyze System.IO.Pipelines, and assuming it is clean, we mark it as "IsTrimmable".

> like this: System.Private.DataContractSerialization.csproj

That one is part of the shared fx, so it gets trimmed today 🙃.

marek-safar replied:
We mark all non-OOB packages are trimmable today and in 5.0 and they have hundreds of linker warnings. Is there any reason why we need to handle OOB packages differently than non-OOB packages?

ViktorHofer replied:
AFAIK OOBs are currently not trimmed.

eerhardt replied:
I don't think the process we used in 5.0 was ideal. We didn't have the tools to do this the "right" way then. But now we do. 

We got a few bugs in 5.0 where someone would use a library no one had tested, and reported issues that it flat out didn't work once trimmed.

marek-safar replied:
> We got a few bugs in 5.0 where someone would use a library no one had tested, and reported issues that it flat out didn't work once trimmed.

Right and I think it's better to get that feedback during early previews where we can priorities better.

Relying on trimming warnings won't help because enabling trimming for libraryA can still break it, if libraryB is not trimming safe and assumes for example some attributes layout, internal members, etc

sbomer replied:
Per discussion with @eerhardt, the plan is to opt all assemblies in as suggested by @marek-safar - and there's a new condition that we can use to opt out per project, should we need to.

By following the existing pattern for AssemblyMetadata like "Serviceable" and "PreferInbox", the attribute will be placed in src and ref assemblies, and also in assemblies targeting netstandard/netcoreapp/netfx. We need to include it for netstandard/netcoreapp because plenty of OOB assemblies build for those targets. Including the attributes in netfx assemblies shouldn't cause any issues since there are no common scenarios where an app targeting net6.0 references one of our netfx assemblies. I think it makes sense to be consistent with existing attributes on this, but @marek-safar if you prefer I could also disable it for netfx.

safern replied:
> and ref assemblies

Are ref assemblies ever trimmed?

marek-safar replied:
> I think it makes sense to be consistent with existing attributes on this, but @marek-safar if you prefer I could also disable it for netfx.

I'm not against it, just wanted to be more cautious for netfx to not break bad code which could, for example, read only the last AssemblyMetadata but the chance is low.

eerhardt replied:
> Are ref assemblies ever trimmed?

No. But they are also never "serviced", and yet we are marking them with "Serviceable" assembly metadata. We should consistently apply these attributes across the assemblies.

safern replied:
> No. But they are also never "serviced", and yet we are marking them with "Serviceable" assembly metadata. We should consistently apply these attributes across the assemblies.

FWIW starting .NET 5, we can now service the ref pack, we try to avoid it but we added support to be able to do it in case we need it, but I'm fine with adding IsTrimmable to them for consistency.

eerhardt commented:
Will this resolve https://github.com/dotnet/aspnetcore/issues/27782 by itself? Or will we need more work to fix that?

sbomer commented:
If we opt System.IO.Pipelines in as part of the change, there shouldn't be any more work in runtime or aspnet, but we'll need to wait for https://github.com/mono/linker/pull/1839 to flow, and for the upcoming SDK change that will include an update to use the new options.

eerhardt commented on eng/versioning.targets:
```diff
@@ -22,6 +22,22 @@
+  <!-- Assembly metadata indicating that an assembly should be trimmed by ILLink (during a self-contained publish - this
+       is independent of whether we link during the runtime build). May be set by test projects.
+       The default is to include this metadata for netcoreapp assemblies but not OOB assemblies. -->
+  <PropertyGroup Condition="'$(IsTestProject)' != 'true' and '$(IsTestSupportProject)' != 'true' and ('$(IsNETCoreAppRef)' == 'true' or '$(IsNETCoreAppSrc)' == 'true')">
```
Can this just be?

```suggestion
  <PropertyGroup Condition="'$(IsNETCoreAppSrc)' == 'true'">
```

sbomer replied:
N/A after recent change

eerhardt commented on eng/versioning.targets:
```diff
@@ -22,6 +22,22 @@
+    <AssemblyMetadata Include="IsTrimmable"
+                      Condition="'$(SetIsTrimmable)' == 'true' and
+                                '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and
+                                $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '6.0'))">
```
This isn't going to work. We build assemblies for netstandard that we will want to make trimmable.

sbomer replied:
N/A (check was removed)

eerhardt commented on src/libraries/System.IO.Pipelines/src/System.IO.Pipelines.csproj:
```diff
@@ -3,6 +3,7 @@
     <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;netcoreapp3.0;net461</TargetFrameworks>
     <ExcludeCurrentNetCoreAppFromPackage>true</ExcludeCurrentNetCoreAppFromPackage>
     <Nullable>enable</Nullable>
+    <ILLinkTrimAssembly>true</ILLinkTrimAssembly>
```
What's the intention here? This runs the linker during the build to trim code. But it doesn't fail on warnings.

sbomer replied:
N/A as discussed - good catch though, thanks.

sbomer committed:
Add a condition to let projects opt out

A project just needs to set <SetIsTrimmable>false</SetIsTrimmable>, and
it won't get the trimmable metadata.

I also got rid of the SkipFrameworkAssemblyMetadata check since I don't
think the IsTrimmable metadata should be tied to whether we are building
a framework assembly, and this property was unused as far as I could tell.

eerhardt reviewed (approved):
Thanks @sbomer.

Note, after this gets plumbed all the way through, we should be able to remove the `StartsWith('Microsoft.Extensions.')` workaround in the blazor SDK (cc @pranavkm @captainsafia)

https://github.com/dotnet/sdk/blob/7029c8c218d113e69db39e69aae5a53efd26f244/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets#L437-L439

safern commented on eng/versioning.targets:
```diff
@@ -83,7 +86,7 @@
       </AssemblyAttribute>
 
       <AssemblyAttribute Include="System.Reflection.AssemblyMetadata"
-                         Condition="'$(SkipFrameworkAssemblyMetadata)' != 'true' and '@(AssemblyMetadata)' != ''">
```
Just curious, why are we removing the `SkipFrameworkAssemblyMetadata` condition?

eerhardt replied:
It doesn't appear to be used.

https://github.com/dotnet/runtime/search?q=SkipFrameworkAssemblyMetadata

Also, I'm assuming it is for the case where `IsTrimmable` should be generated for the assembly, but someone adds `SkipFrameworkAssemblyMetadata` to the project. That would block `IsTrimmable` from getting generated.

sbomer replied:
Yup - see also the second commit message: https://github.com/dotnet/runtime/pull/48428/commits/d03327e62ddb7035442c9b803c699d1f8776e1bd

stephentoub commented on src/libraries/System.Private.CoreLib/src/Internal/AssemblyAttributes.cs:
```diff
@@ -22,5 +22,6 @@
 
 [assembly: AssemblyMetadata("Serviceable", "True")]
 [assembly: AssemblyMetadata(".NETFrameworkAssembly", "")]
```
Completely unrelated to this PR, but it caught my eye... is this value still desired?  Should the "Framework" piece be removed?
cc: @ericstj

eerhardt replied:
Certain things look for this string. ex.

https://github.com/dotnet/corert/blob/24f1edc938af9954eac9a3385958787ca64956d5/src/ILCompiler.Compiler/src/Compiler/FrameworkStringResourceBlockingPolicy.cs#L58-L87
